copycat/runtime/src/main/java/org/apache/kafka/copycat/runtime/distributed/KafkaConfigStorage.java

```diff
@@ -0,0 +1,438 @@
+        }
+
+        // Start sending all the individual updates
+        for (Map.Entry<ConnectorTaskId, Map<String, String>> taskConfigEntry : configs.entrySet()) {
+            Struct copycatConfig = new Struct(TASK_CONFIGURATION_V0);
+            copycatConfig.put("properties", taskConfigEntry.getValue());
+            byte[] serializedConfig = converter.fromCopycatData(topic, TASK_CONFIGURATION_V0, copycatConfig);
+            configLog.send(TASK_KEY(taskConfigEntry.getKey()), serializedConfig);
+        }
+
+        // Finally, send the root update, then wait until we read to the end of the log
+        try {
+            // Read to end to ensure all the task configs have been written
+            configLog.readToEnd().get(READ_TO_END_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+
+            Struct copycatConfig = new Struct(ROOT_V0);
+            copycatConfig.put("connectors", newRootConfig);
+            byte[] serializedConfig = converter.fromCopycatData(topic, ROOT_V0, copycatConfig);
+            configLog.send(ROOT_KEY, serializedConfig);
+
+            configLog.readToEnd().get(READ_TO_END_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+        } catch (InterruptedException | ExecutionException | TimeoutException e) {
+            log.error("Failed to write root configuration to Kafka: ", e);
+            throw new CopycatException("Error writing root configuration to Kafka", e);
+        }
+    }
+
+    private KafkaBasedLog<String, byte[]> createKafkaBasedLog(String topic, Map<String, Object> producerProps,
+                                                              Map<String, Object> consumerProps, Callback<ConsumerRecord<String, byte[]>> consumedCallback) {
+        return new KafkaBasedLog<>(topic, producerProps, consumerProps, consumedCallback, new SystemTime());
+    }
+
+    private final Callback<ConsumerRecord<String, byte[]>> consumedCallback = new Callback<ConsumerRecord<String, byte[]>>() {
+        @Override
+        public void onCompletion(Throwable error, ConsumerRecord<String, byte[]> record) {
+            if (error != null) {
+                log.error("Unexpected in consumer callback for KafkaConfigStorage: ", error);
+                return;
+            }
+
+            final SchemaAndValue value;
+            try {
+                value = converter.toCopycatData(topic, record.value());
+            } catch (DataException e) {
+                log.error("Failed to convert config data to Copycat format: ", e);
+                return;
+            }
+
+            if (record.key().equals(ROOT_KEY)) {
+                List<ConnectorTaskId> updatedTasks;
+                synchronized (lock) {
+                    // Apply any outstanding deferred task updates and then update the root config. This *can* result in
+                    // an inconsistent output, which will have to be resolved by some external mechanism ensuring reconfiguration
```

Comment:
Mind documenting the explicit case in the class? Right now the warnings about inconsistencies are rather abstract and it will help people who touch the code to know what we are trying to guard against. This will help in keeping the code correct in the future, as we will have something concrete to reason about.